pkg/auditlog/auditlog.go

```diff
@@ -0,0 +1,88 @@
+// Package auditlog is a package to send audit logs events to WorkOS.
+//
+// Example:
+//   func main() {
+//       auditlog.SetAPIKey("my_api_key")
+//       defer auditlog.Close()
+//
+//       // Wherever you need to publish an audit log event:
+//       auditlog.Publish(auditlog.Event{
+//           Action:     "document.viewed",
+//           ActionType: "r",
+//           ActorName:  "Jairo Kunde",
+//           ActorID:    "user_01DGZ0FAXN978HCET66Q98QMTQ",
+//           Group:      "abstract.com",
+//           Location:   "55.27.223.26",
+//           OccurredAt: time.Now(),
+//           TargetName: "central.class",
+//           TargetID:   "doc_01DGZ0FAXP4HA4X0BVFKS0ZH4Y",
+//       })
+//   }
+package auditlog
+
+import (
+	"os"
+	"time"
+)
+
+var (
+	// DefaultPublisher is the publisher used by SetAPIKey, Publish and Close
+	// functions.
+	DefaultPublisher = &Publisher{
+		APIKey:    os.Getenv("WORKOS_API_KEY"),
+		Endpoint:  "https://api.workos.com/events",
+		QueueSize: 512,
+	}
+
+	currentLocation string
+)
+
+func init() {
+	currentLocation, _ = os.Hostname()
+}
+
+// SetAPIKey sets the WorkOS API key to use when using Publish.
+func SetAPIKey(k string) {
+	DefaultPublisher.APIKey = k
+}
+
+// Publish publishes the given events.
+func Publish(events ...Event) {
+	DefaultPublisher.Publish(events...)
+}
+
+// Close stops publishings audit log events and releases allocated resources.
+// It waits for pending events to be sent before returning.
+func Close() {
+	DefaultPublisher.Close()
+}
+
+// Event represents an Audit Log event.
+type Event struct {
+	Action     string                 `json:"action"`
+	ActionType string                 `json:"action_type"`
+	ActorName  string                 `json:"actor_name"`
+	ActorID    string                 `json:"actor_id"`
+	Group      string                 `json:"group"`
+	Location   string                 `json:"location"`
+	Metadata   map[string]interface{} `json:"metadata,omitempty"`
```

Comment:
I forgot to implement the global metadata, I ll add it to the implementation.
I'm curious about the metadata validation.

Reply:
```One thing I would have added if I took another stab at it would be making the metadata field not exported and only mutable by exported functions. This would give us more control of the types of values the client would know they could attach to it.```

It could be a solution. Is it possible to have multiple level in metadata?
```go
meta := map[string]interface{}{
    "simple": 42,
    "nested": map[string]interface{}{
        "nested simple": 42,
    },
    "struct": struct{
        Field string
    }{
        Field: "hello", 
   },
}
```

Reply:
> I'm curious about the metadata validation.

The current implementation handles the API specs last I saw them, but @marktran and @maxchehab can chime in further if those are still accurate.

> It could be a solution. Is it possible to have multiple level in metadata?

I believe the answer is no, but @marktran and @maxchehab answer with more clarity.

Reply:
Another question:
In the current global metadata implementation, in case of duplicate key, look like the event metadata will always be overridden by the global one:
```go
	// Add the global metadata to the Event's metadata
	for k, v := range globalMetadata {
		e.Metadata[k] = v
	}
```

Is it the behavior we want?
The one set at the event level are the one that are the one who are the most up to date. Have the feeling that they are the one that should prevail.

Reply:
An additional check can be performed here to see if the key is already set or the local metadata can override the global metadata. It's definitely a good question with no easy answer.

Reply:
@maxence-charriere We talked about having a metadata limit similar to Stripe:

> You can specify up to 50 keys, with key names up to 40 characters long and values up to 500 characters long.
\- https://stripe.com/docs/api/metadata

This isn't enforced consistently at the moment.

Reply:
This could be documented. I did not see that those check are done in the js sdk. do we want to throw an error if a key is longer than 50 character or do we want to trim the key? How this is handled in js?

Reply:
Added a story internally for this.